File: customer-service/src/main/java/com/wks/servicemarketplace/customerservice/core/utils/ModelValidator.java
package com.wks.servicemarketplace.customerservice.core.utils;

import com.wks.servicemarketplace.customerservice.core.exceptions.ValidationException;

import javax.validation.ConstraintViolation;
import javax.validation.Validation;
import javax.validation.Validator;
import javax.validation.ValidatorFactory;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

public class ModelValidator {

    private static final Validator validator;

    static {
        ValidatorFactory factory = Validation.buildDefaultValidatorFactory();
        validator = factory.getValidator();
    }

    public static <T> T validate(T instance, Class<?>... clazz) {
        final Set<ConstraintViolation<T>> violations = validator.validate(instance, clazz);
        if (!violations.isEmpty()) {
            final Map<String, String> fields = violations.stream()
                    .collect(Collectors.toMap(it -> it.getPropertyPath().toString(), ConstraintViolation::getMessage));
            throw new ValidationException(fields);
        }
        return instance;
    }
}

File: service-provider-service/src/main/kotlin/com/wks/servicemarketplace/serviceproviderservice/core/utils/ModelValidator.kt
package com.wks.servicemarketplace.serviceproviderservice.core.utils

import com.wks.servicemarketplace.serviceproviderservice.core.exceptions.ValidationException
import javax.validation.Validation

object ModelValidator {
    private var validator = Validation.buildDefaultValidatorFactory().validator

    fun <T> validate(instance: T, vararg clazz: Class<*>?): T {
        val violations = validator.validate(instance, *clazz)
        if (violations.isNotEmpty()) {
            val fields = violations.map { it.propertyPath.toString() to listOf(it.message) }.toMap()
            throw ValidationException(fields)
        }
        return instance
    }

}